// @0x/abi-gen-wrappers/lib/src/generated-wrappers/dev_utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:no-consecutive-blank-lines ordered-imports align trailing-comma
// tslint:disable:whitespace no-unbound-method no-trailing-whitespace
// tslint:disable:no-unused-variable
var base_contract_1 = require("@0x/base-contract");
var json_schemas_1 = require("@0x/json-schemas");
var utils_1 = require("@0x/utils");
var web3_wrapper_1 = require("@0x/web3-wrapper");
var assert_1 = require("@0x/assert");
var ethers = require("ethers");
// tslint:enable:no-unused-variable
/* istanbul ignore next */
// tslint:disable:no-parameter-reassignment
// tslint:disable-next-line:class-name
var DevUtilsContract = /** @class */ (function (_super) {
    __extends(DevUtilsContract, _super);
    function DevUtilsContract(address, supportedProvider, txDefaults, logDecodeDependencies) {
        var _this = _super.call(this, 'DevUtils', DevUtilsContract.ABI(), address, supportedProvider, txDefaults, logDecodeDependencies) || this;
        /**
         * Decode ERC-721 asset data from the format described in the AssetProxy contract specification.
         */
        _this.decodeERC721AssetData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param assetData AssetProxy-compliant asset data describing an ERC-721
             *     asset.
             * @returns The ERC-721 AssetProxy identifier, the address of the ERC-721 contract hosting this asset, and the identifier of the specific asset to be traded.
             */
            callAsync: function (assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('decodeERC721AssetData(bytes)', [assetData]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('decodeERC721AssetData(bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param assetData AssetProxy-compliant asset data describing an ERC-721
             *     asset.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (assetData) {
                assert_1.assert.isString('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('decodeERC721AssetData(bytes)', [assetData]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeERC721AssetData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeERC721AssetData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Calls getBalance() and getAllowance() for assetData.
         */
        _this.getBalanceAndAssetProxyAllowance = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Details of asset, encoded per the AssetProxy contract
             *     specification.
             * @returns Number of assets (or asset baskets) held by owner, and number of assets (or asset baskets) that the corresponding AssetProxy is authorized to spend.
             */
            callAsync: function (ownerAddress, assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('ownerAddress', ownerAddress);
                                assert_1.assert.isString('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getBalanceAndAssetProxyAllowance(address,bytes)', [
                                    ownerAddress.toLowerCase(),
                                    assetData,
                                ]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getBalanceAndAssetProxyAllowance(address,bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Details of asset, encoded per the AssetProxy contract
             *     specification.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (ownerAddress, assetData) {
                assert_1.assert.isString('ownerAddress', ownerAddress);
                assert_1.assert.isString('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getBalanceAndAssetProxyAllowance(address,bytes)', [ownerAddress.toLowerCase(), assetData]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBalanceAndAssetProxyAllowance(address,bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBalanceAndAssetProxyAllowance(address,bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        _this.ERC1155_PROXY_ID = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             */
            callAsync: function (callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('ERC1155_PROXY_ID()', []);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('ERC1155_PROXY_ID()');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function () {
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('ERC1155_PROXY_ID()', []);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('ERC1155_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('ERC1155_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Gets the amount of an asset transferable by the owner.
         */
        _this.getTransferableAssetAmount = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param ownerAddress Address of the owner of the asset.
             * @param assetData Description of tokens, per the AssetProxy contract
             *     specification.
             * @returns The amount of the asset tranferable by the owner. NOTE: If the &#x60;assetData&#x60; encodes data for multiple assets, the &#x60;transferableAssetAmount&#x60; will represent the amount of times the entire &#x60;assetData&#x60; can be transferred. To calculate the total individual transferable amounts, this scaled &#x60;transferableAmount&#x60; must be multiplied by  the individual asset amounts located within the &#x60;assetData&#x60;.
             */
            callAsync: function (ownerAddress, assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('ownerAddress', ownerAddress);
                                assert_1.assert.isString('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getTransferableAssetAmount(address,bytes)', [
                                    ownerAddress.toLowerCase(),
                                    assetData,
                                ]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getTransferableAssetAmount(address,bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param ownerAddress Address of the owner of the asset.
             * @param assetData Description of tokens, per the AssetProxy contract
             *     specification.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (ownerAddress, assetData) {
                assert_1.assert.isString('ownerAddress', ownerAddress);
                assert_1.assert.isString('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getTransferableAssetAmount(address,bytes)', [
                    ownerAddress.toLowerCase(),
                    assetData,
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getTransferableAssetAmount(address,bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getTransferableAssetAmount(address,bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Calls getAssetProxyAllowance() for each element of assetData.
         */
        _this.getBatchAssetProxyAllowances = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Array of asset details, each encoded per the AssetProxy
             *     contract specification.
             * @returns An array of asset allowances from getAllowance(), with each element corresponding to the same-indexed element in the assetData input.
             */
            callAsync: function (ownerAddress, assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('ownerAddress', ownerAddress);
                                assert_1.assert.isArray('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getBatchAssetProxyAllowances(address,bytes[])', [
                                    ownerAddress.toLowerCase(),
                                    assetData,
                                ]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getBatchAssetProxyAllowances(address,bytes[])');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Array of asset details, each encoded per the AssetProxy
             *     contract specification.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (ownerAddress, assetData) {
                assert_1.assert.isString('ownerAddress', ownerAddress);
                assert_1.assert.isArray('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getBatchAssetProxyAllowances(address,bytes[])', [ownerAddress.toLowerCase(), assetData]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBatchAssetProxyAllowances(address,bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBatchAssetProxyAllowances(address,bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Encode ERC-20 asset data into the format described in the AssetProxy contract specification.
         */
        _this.encodeERC20AssetData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param tokenAddress The address of the ERC-20 contract hosting the asset to
             *     be traded.
             * @returns AssetProxy-compliant data describing the asset.
             */
            callAsync: function (tokenAddress, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('tokenAddress', tokenAddress);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('encodeERC20AssetData(address)', [
                                    tokenAddress.toLowerCase(),
                                ]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('encodeERC20AssetData(address)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param tokenAddress The address of the ERC-20 contract hosting the asset to
             *     be traded.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (tokenAddress) {
                assert_1.assert.isString('tokenAddress', tokenAddress);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('encodeERC20AssetData(address)', [
                    tokenAddress.toLowerCase(),
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('encodeERC20AssetData(address)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('encodeERC20AssetData(address)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Decodes the call data for an Exchange contract method call.
         */
        _this.decodeZeroExTransactionData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param transactionData ABI-encoded calldata for an Exchange     contract
             *     method call.
             * @returns The name of the function called, and the parameters it was     given.  For single-order fills and cancels, the arrays will have     just one element.
             */
            callAsync: function (transactionData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('transactionData', transactionData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('decodeZeroExTransactionData(bytes)', [transactionData]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('decodeZeroExTransactionData(bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param transactionData ABI-encoded calldata for an Exchange     contract
             *     method call.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (transactionData) {
                assert_1.assert.isString('transactionData', transactionData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('decodeZeroExTransactionData(bytes)', [
                    transactionData,
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeZeroExTransactionData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeZeroExTransactionData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Returns the owner's balance of the assets(s) specified in assetData.  When the asset data contains multiple assets (eg in ERC1155 or Multi-Asset), the return value indicates how many complete "baskets" of those assets are owned by owner.
         */
        _this.getBalance = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Details of asset, encoded per the AssetProxy contract
             *     specification.
             * @returns Number of assets (or asset baskets) held by owner.
             */
            callAsync: function (ownerAddress, assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('ownerAddress', ownerAddress);
                                assert_1.assert.isString('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getBalance(address,bytes)', [
                                    ownerAddress.toLowerCase(),
                                    assetData,
                                ]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getBalance(address,bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Details of asset, encoded per the AssetProxy contract
             *     specification.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (ownerAddress, assetData) {
                assert_1.assert.isString('ownerAddress', ownerAddress);
                assert_1.assert.isString('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getBalance(address,bytes)', [
                    ownerAddress.toLowerCase(),
                    assetData,
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBalance(address,bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBalance(address,bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Fetches all order-relevant information needed to validate if the supplied orders are fillable.
         */
        _this.getOrderRelevantStates = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param orders Array of order structures.
             * @param signatures Array of signatures provided by makers that prove the
             *     authenticity of the orders. `0x01` can always be provided if a signature
             *     does not need to be validated.
             * @returns The ordersInfo (array of the hash, status, and &#x60;takerAssetAmount&#x60; already filled for each order), fillableTakerAssetAmounts (array of amounts for each order&#x27;s &#x60;takerAssetAmount&#x60; that is fillable given all on-chain state), and isValidSignature (array containing the validity of each provided signature). NOTE: If the &#x60;takerAssetData&#x60; encodes data for multiple assets, each element of &#x60;fillableTakerAssetAmounts&#x60; will represent a &quot;scaled&quot; amount, meaning it must be multiplied by all the individual asset amounts within the &#x60;takerAssetData&#x60; to get the final amount of each asset that can be filled.
             */
            callAsync: function (orders, signatures, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isArray('orders', orders);
                                assert_1.assert.isArray('signatures', signatures);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getOrderRelevantStates((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes)[],bytes[])', [orders, signatures]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getOrderRelevantStates((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes)[],bytes[])');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param orders Array of order structures.
             * @param signatures Array of signatures provided by makers that prove the
             *     authenticity of the orders. `0x01` can always be provided if a signature
             *     does not need to be validated.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (orders, signatures) {
                assert_1.assert.isArray('orders', orders);
                assert_1.assert.isArray('signatures', signatures);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getOrderRelevantStates((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes)[],bytes[])', [orders, signatures]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getOrderRelevantStates((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes)[],bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getOrderRelevantStates((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes)[],bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        _this.ERC20_PROXY_ID = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             */
            callAsync: function (callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('ERC20_PROXY_ID()', []);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('ERC20_PROXY_ID()');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function () {
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('ERC20_PROXY_ID()', []);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('ERC20_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('ERC20_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Decode ERC-20 asset data from the format described in the AssetProxy contract specification.
         */
        _this.decodeERC20AssetData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param assetData AssetProxy-compliant asset data describing an ERC-20 asset.
             * @returns The ERC-20 AssetProxy identifier, and the address of the ERC-20  contract hosting this asset.
             */
            callAsync: function (assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('decodeERC20AssetData(bytes)', [assetData]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('decodeERC20AssetData(bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param assetData AssetProxy-compliant asset data describing an ERC-20 asset.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (assetData) {
                assert_1.assert.isString('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('decodeERC20AssetData(bytes)', [assetData]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeERC20AssetData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeERC20AssetData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Fetches all order-relevant information needed to validate if the supplied order is fillable.
         */
        _this.getOrderRelevantState = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param order The order structure.
             * @param signature Signature provided by maker that proves the order's
             *     authenticity. `0x01` can always be provided if the signature does not
             *     need to be validated.
             * @returns The orderInfo (hash, status, and &#x60;takerAssetAmount&#x60; already filled for the given order), fillableTakerAssetAmount (amount of the order&#x27;s &#x60;takerAssetAmount&#x60; that is fillable given all on-chain state), and isValidSignature (validity of the provided signature). NOTE: If the &#x60;takerAssetData&#x60; encodes data for multiple assets, &#x60;fillableTakerAssetAmount&#x60; will represent a &quot;scaled&quot; amount, meaning it must be multiplied by all the individual asset amounts within the &#x60;takerAssetData&#x60; to get the final amount of each asset that can be filled.
             */
            callAsync: function (order, signature, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('signature', signature);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getOrderRelevantState((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes),bytes)', [order, signature]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getOrderRelevantState((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes),bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param order The order structure.
             * @param signature Signature provided by maker that proves the order's
             *     authenticity. `0x01` can always be provided if the signature does not
             *     need to be validated.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (order, signature) {
                assert_1.assert.isString('signature', signature);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getOrderRelevantState((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes),bytes)', [order, signature]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getOrderRelevantState((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes),bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getOrderRelevantState((address,address,address,address,uint256,uint256,uint256,uint256,uint256,uint256,bytes,bytes),bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Decode ERC-1155 asset data from the format described in the AssetProxy contract specification.
         */
        _this.decodeERC1155AssetData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param assetData AssetProxy-compliant asset data describing an ERC-1155 set
             *     of assets.
             * @returns The ERC-1155 AssetProxy identifier, the address of the ERC-1155 contract hosting the assets, an array of the identifiers of the assets to be traded, an array of asset amounts to be traded, and callback data.  Each element of the arrays corresponds to the same-indexed element of the other array.  Return values specified as &#x60;memory&#x60; are returned as pointers to locations within the memory of the input parameter &#x60;assetData&#x60;.
             */
            callAsync: function (assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('decodeERC1155AssetData(bytes)', [assetData]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('decodeERC1155AssetData(bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param assetData AssetProxy-compliant asset data describing an ERC-1155 set
             *     of assets.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (assetData) {
                assert_1.assert.isString('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('decodeERC1155AssetData(bytes)', [assetData]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeERC1155AssetData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeERC1155AssetData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Batch fetches ETH balances
         */
        _this.getEthBalances = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param addresses Array of addresses.
             * @returns Array of ETH balances.
             */
            callAsync: function (addresses, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isArray('addresses', addresses);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getEthBalances(address[])', [addresses]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getEthBalances(address[])');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param addresses Array of addresses.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (addresses) {
                assert_1.assert.isArray('addresses', addresses);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getEthBalances(address[])', [addresses]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getEthBalances(address[])');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getEthBalances(address[])');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        _this.ERC721_PROXY_ID = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             */
            callAsync: function (callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('ERC721_PROXY_ID()', []);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('ERC721_PROXY_ID()');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function () {
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('ERC721_PROXY_ID()', []);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('ERC721_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('ERC721_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Encode ERC-721 asset data into the format described in the AssetProxy specification.
         */
        _this.encodeERC721AssetData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param tokenAddress The address of the ERC-721 contract hosting the asset to
             *     be traded.
             * @param tokenId The identifier of the specific asset to be traded.
             * @returns AssetProxy-compliant asset data describing the asset.
             */
            callAsync: function (tokenAddress, tokenId, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('tokenAddress', tokenAddress);
                                assert_1.assert.isBigNumber('tokenId', tokenId);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('encodeERC721AssetData(address,uint256)', [
                                    tokenAddress.toLowerCase(),
                                    tokenId,
                                ]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('encodeERC721AssetData(address,uint256)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param tokenAddress The address of the ERC-721 contract hosting the asset to
             *     be traded.
             * @param tokenId The identifier of the specific asset to be traded.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (tokenAddress, tokenId) {
                assert_1.assert.isString('tokenAddress', tokenAddress);
                assert_1.assert.isBigNumber('tokenId', tokenId);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('encodeERC721AssetData(address,uint256)', [
                    tokenAddress.toLowerCase(),
                    tokenId,
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('encodeERC721AssetData(address,uint256)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('encodeERC721AssetData(address,uint256)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        _this.MULTI_ASSET_PROXY_ID = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             */
            callAsync: function (callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('MULTI_ASSET_PROXY_ID()', []);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('MULTI_ASSET_PROXY_ID()');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function () {
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('MULTI_ASSET_PROXY_ID()', []);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('MULTI_ASSET_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('MULTI_ASSET_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Encode ERC-1155 asset data into the format described in the AssetProxy contract specification.
         */
        _this.encodeERC1155AssetData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param tokenAddress The address of the ERC-1155 contract hosting the
             *     asset(s) to be traded.
             * @param tokenIds The identifiers of the specific assets to be traded.
             * @param tokenValues The amounts of each asset to be traded.
             * @param callbackData Data to be passed to receiving contracts when a transfer
             *     is performed.
             * @returns AssetProxy-compliant asset data describing the set of assets.
             */
            callAsync: function (tokenAddress, tokenIds, tokenValues, callbackData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('tokenAddress', tokenAddress);
                                assert_1.assert.isArray('tokenIds', tokenIds);
                                assert_1.assert.isArray('tokenValues', tokenValues);
                                assert_1.assert.isString('callbackData', callbackData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('encodeERC1155AssetData(address,uint256[],uint256[],bytes)', [tokenAddress.toLowerCase(), tokenIds, tokenValues, callbackData]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('encodeERC1155AssetData(address,uint256[],uint256[],bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param tokenAddress The address of the ERC-1155 contract hosting the
             *     asset(s) to be traded.
             * @param tokenIds The identifiers of the specific assets to be traded.
             * @param tokenValues The amounts of each asset to be traded.
             * @param callbackData Data to be passed to receiving contracts when a transfer
             *     is performed.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (tokenAddress, tokenIds, tokenValues, callbackData) {
                assert_1.assert.isString('tokenAddress', tokenAddress);
                assert_1.assert.isArray('tokenIds', tokenIds);
                assert_1.assert.isArray('tokenValues', tokenValues);
                assert_1.assert.isString('callbackData', callbackData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('encodeERC1155AssetData(address,uint256[],uint256[],bytes)', [tokenAddress.toLowerCase(), tokenIds, tokenValues, callbackData]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('encodeERC1155AssetData(address,uint256[],uint256[],bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('encodeERC1155AssetData(address,uint256[],uint256[],bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Calls `asset.ownerOf(tokenId)`, but returns a null owner instead of reverting on an unowned asset.
         */
        _this.getERC721TokenOwner = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param tokenAddress Address of ERC721 asset.
             * @param tokenId The identifier for the specific NFT.
             * @returns Owner of tokenId or null address if unowned.
             */
            callAsync: function (tokenAddress, tokenId, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('tokenAddress', tokenAddress);
                                assert_1.assert.isBigNumber('tokenId', tokenId);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getERC721TokenOwner(address,uint256)', [
                                    tokenAddress.toLowerCase(),
                                    tokenId,
                                ]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getERC721TokenOwner(address,uint256)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param tokenAddress Address of ERC721 asset.
             * @param tokenId The identifier for the specific NFT.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (tokenAddress, tokenId) {
                assert_1.assert.isString('tokenAddress', tokenAddress);
                assert_1.assert.isBigNumber('tokenId', tokenId);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getERC721TokenOwner(address,uint256)', [
                    tokenAddress.toLowerCase(),
                    tokenId,
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getERC721TokenOwner(address,uint256)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getERC721TokenOwner(address,uint256)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Decode multi-asset data from the format described in the AssetProxy contract specification.
         */
        _this.decodeMultiAssetData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param assetData AssetProxy-compliant data describing a multi-asset basket.
             * @returns The Multi-Asset AssetProxy identifier, an array of the amounts of the assets to be traded, and an array of the AssetProxy-compliant data describing each asset to be traded.  Each element of the arrays corresponds to the same-indexed element of the other array.
             */
            callAsync: function (assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('decodeMultiAssetData(bytes)', [assetData]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('decodeMultiAssetData(bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param assetData AssetProxy-compliant data describing a multi-asset basket.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (assetData) {
                assert_1.assert.isString('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('decodeMultiAssetData(bytes)', [assetData]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeMultiAssetData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('decodeMultiAssetData(bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Calls getBalance() for each element of assetData.
         */
        _this.getBatchBalances = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Array of asset details, each encoded per the AssetProxy
             *     contract specification.
             * @returns Array of asset balances from getBalance(), with each element corresponding to the same-indexed element in the assetData input.
             */
            callAsync: function (ownerAddress, assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('ownerAddress', ownerAddress);
                                assert_1.assert.isArray('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getBatchBalances(address,bytes[])', [
                                    ownerAddress.toLowerCase(),
                                    assetData,
                                ]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getBatchBalances(address,bytes[])');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Array of asset details, each encoded per the AssetProxy
             *     contract specification.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (ownerAddress, assetData) {
                assert_1.assert.isString('ownerAddress', ownerAddress);
                assert_1.assert.isArray('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getBatchBalances(address,bytes[])', [
                    ownerAddress.toLowerCase(),
                    assetData,
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBatchBalances(address,bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBatchBalances(address,bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Returns the number of asset(s) (described by assetData) that the corresponding AssetProxy contract is authorized to spend.  When the asset data contains multiple assets (eg for Multi-Asset), the return value indicates how many complete "baskets" of those assets may be spent by all of the corresponding AssetProxy contracts.
         */
        _this.getAssetProxyAllowance = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Details of asset, encoded per the AssetProxy contract
             *     specification.
             * @returns Number of assets (or asset baskets) that the corresponding AssetProxy is authorized to spend.
             */
            callAsync: function (ownerAddress, assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('ownerAddress', ownerAddress);
                                assert_1.assert.isString('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getAssetProxyAllowance(address,bytes)', [
                                    ownerAddress.toLowerCase(),
                                    assetData,
                                ]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getAssetProxyAllowance(address,bytes)');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Details of asset, encoded per the AssetProxy contract
             *     specification.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (ownerAddress, assetData) {
                assert_1.assert.isString('ownerAddress', ownerAddress);
                assert_1.assert.isString('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getAssetProxyAllowance(address,bytes)', [
                    ownerAddress.toLowerCase(),
                    assetData,
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getAssetProxyAllowance(address,bytes)');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getAssetProxyAllowance(address,bytes)');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Encode data for multiple assets, per the AssetProxy contract specification.
         */
        _this.encodeMultiAssetData = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param amounts The amounts of each asset to be traded.
             * @param nestedAssetData AssetProxy-compliant data describing each asset to be
             *     traded.
             * @returns AssetProxy-compliant data describing the set of assets.
             */
            callAsync: function (amounts, nestedAssetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, encodedDataBytes, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isArray('amounts', amounts);
                                assert_1.assert.isArray('nestedAssetData', nestedAssetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('encodeMultiAssetData(uint256[],bytes[])', [
                                    amounts,
                                    nestedAssetData,
                                ]);
                                encodedDataBytes = Buffer.from(encodedData.substr(2), 'hex');
                                return [4 /*yield*/, self.evmExecAsync(encodedDataBytes)];
                            case 1:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('encodeMultiAssetData(uint256[],bytes[])');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param amounts The amounts of each asset to be traded.
             * @param nestedAssetData AssetProxy-compliant data describing each asset to be
             *     traded.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (amounts, nestedAssetData) {
                assert_1.assert.isArray('amounts', amounts);
                assert_1.assert.isArray('nestedAssetData', nestedAssetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('encodeMultiAssetData(uint256[],bytes[])', [
                    amounts,
                    nestedAssetData,
                ]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('encodeMultiAssetData(uint256[],bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('encodeMultiAssetData(uint256[],bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        _this.STATIC_CALL_PROXY_ID = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             */
            callAsync: function (callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('STATIC_CALL_PROXY_ID()', []);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('STATIC_CALL_PROXY_ID()');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function () {
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('STATIC_CALL_PROXY_ID()', []);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('STATIC_CALL_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('STATIC_CALL_PROXY_ID()');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        /**
         * Calls getBatchBalances() and getBatchAllowances() for each element of assetData.
         */
        _this.getBatchBalancesAndAssetProxyAllowances = {
            /**
             * Sends a read-only call to the contract method. Returns the result that would happen if one were to send an
             * Ethereum transaction to this method, given the current state of the blockchain. Calls do not cost gas
             * since they don't modify state.
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Array of asset details, each encoded per the AssetProxy
             *     contract specification.
             * @returns An array of asset balances from getBalance(), and an array of asset allowances from getAllowance(), with each element corresponding to the same-indexed element in the assetData input.
             */
            callAsync: function (ownerAddress, assetData, callData, defaultBlock) {
                if (callData === void 0) { callData = {}; }
                return __awaiter(this, void 0, void 0, function () {
                    var self, encodedData, callDataWithDefaults, rawCallResult, abiEncoder, result;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                assert_1.assert.isString('ownerAddress', ownerAddress);
                                assert_1.assert.isArray('assetData', assetData);
                                assert_1.assert.doesConformToSchema('callData', callData, json_schemas_1.schemas.callDataSchema, [
                                    json_schemas_1.schemas.addressSchema,
                                    json_schemas_1.schemas.numberSchema,
                                    json_schemas_1.schemas.jsNumber,
                                ]);
                                if (defaultBlock !== undefined) {
                                    assert_1.assert.isBlockParam('defaultBlock', defaultBlock);
                                }
                                self = this;
                                encodedData = self._strictEncodeArguments('getBatchBalancesAndAssetProxyAllowances(address,bytes[])', [ownerAddress.toLowerCase(), assetData]);
                                return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync(__assign({ to: self.address }, callData, { data: encodedData }), self._web3Wrapper.getContractDefaults())];
                            case 1:
                                callDataWithDefaults = _a.sent();
                                callDataWithDefaults.from = callDataWithDefaults.from
                                    ? callDataWithDefaults.from.toLowerCase()
                                    : callDataWithDefaults.from;
                                return [4 /*yield*/, self._web3Wrapper.callAsync(callDataWithDefaults, defaultBlock)];
                            case 2:
                                rawCallResult = _a.sent();
                                base_contract_1.BaseContract._throwIfRevertWithReasonCallResult(rawCallResult);
                                abiEncoder = self._lookupAbiEncoder('getBatchBalancesAndAssetProxyAllowances(address,bytes[])');
                                result = abiEncoder.strictDecodeReturnValue(rawCallResult);
                                // tslint:enable boolean-naming
                                return [2 /*return*/, result];
                        }
                    });
                });
            },
            /**
             * Returns the ABI encoded transaction data needed to send an Ethereum transaction calling this method. Before
             * sending the Ethereum tx, this encoded tx data can first be sent to a separate signing service or can be used
             * to create a 0x transaction (see protocol spec for more details).
             * @param ownerAddress Owner of the assets specified by assetData.
             * @param assetData Array of asset details, each encoded per the AssetProxy
             *     contract specification.
             * @returns The ABI encoded transaction data as a string
             */
            getABIEncodedTransactionData: function (ownerAddress, assetData) {
                assert_1.assert.isString('ownerAddress', ownerAddress);
                assert_1.assert.isArray('assetData', assetData);
                var self = this;
                var abiEncodedTransactionData = self._strictEncodeArguments('getBatchBalancesAndAssetProxyAllowances(address,bytes[])', [ownerAddress.toLowerCase(), assetData]);
                return abiEncodedTransactionData;
            },
            /**
             * Decode the ABI-encoded transaction data into its input arguments
             * @param callData The ABI-encoded transaction data
             * @returns An array representing the input arguments in order. Keynames of nested structs are preserved.
             */
            getABIDecodedTransactionData: function (callData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBatchBalancesAndAssetProxyAllowances(address,bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedCallData = abiEncoder.strictDecode(callData);
                return abiDecodedCallData;
            },
            /**
             * Decode the ABI-encoded return data from a transaction
             * @param returnData the data returned after transaction execution
             * @returns An array representing the output results in order.  Keynames of nested structs are preserved.
             */
            getABIDecodedReturnData: function (returnData) {
                var self = this;
                var abiEncoder = self._lookupAbiEncoder('getBatchBalancesAndAssetProxyAllowances(address,bytes[])');
                // tslint:disable boolean-naming
                var abiDecodedReturnData = abiEncoder.strictDecodeReturnValue(returnData);
                return abiDecodedReturnData;
            },
        };
        utils_1.classUtils.bindAll(_this, ['_abiEncoderByFunctionSignature', 'address', '_web3Wrapper']);
        return _this;
    }
    DevUtilsContract.deployFrom0xArtifactAsync = function (artifact, supportedProvider, txDefaults, logDecodeDependencies, _exchange, _zrxAssetData) {
        return __awaiter(this, void 0, void 0, function () {
            var e_1, _a, provider, bytecode, abi, logDecodeDependenciesAbiOnly, _b, _c, key;
            return __generator(this, function (_d) {
                assert_1.assert.doesConformToSchema('txDefaults', txDefaults, json_schemas_1.schemas.txDataSchema, [
                    json_schemas_1.schemas.addressSchema,
                    json_schemas_1.schemas.numberSchema,
                    json_schemas_1.schemas.jsNumber,
                ]);
                if (artifact.compilerOutput === undefined) {
                    throw new Error('Compiler output not found in the artifact file');
                }
                provider = utils_1.providerUtils.standardizeOrThrow(supportedProvider);
                bytecode = artifact.compilerOutput.evm.bytecode.object;
                abi = artifact.compilerOutput.abi;
                logDecodeDependenciesAbiOnly = {};
                if (Object.keys(logDecodeDependencies) !== undefined) {
                    try {
                        for (_b = __values(Object.keys(logDecodeDependencies)), _c = _b.next(); !_c.done; _c = _b.next()) {
                            key = _c.value;
                            logDecodeDependenciesAbiOnly[key] = logDecodeDependencies[key].compilerOutput.abi;
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                }
                return [2 /*return*/, DevUtilsContract.deployAsync(bytecode, abi, provider, txDefaults, logDecodeDependenciesAbiOnly, _exchange, _zrxAssetData)];
            });
        });
    };
    DevUtilsContract.deployAsync = function (bytecode, abi, supportedProvider, txDefaults, logDecodeDependencies, _exchange, _zrxAssetData) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, provider, constructorAbi, iface, deployInfo, txData, web3Wrapper, txDataWithDefaults, txHash, txReceipt, contractInstance;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        assert_1.assert.isHexString('bytecode', bytecode);
                        assert_1.assert.doesConformToSchema('txDefaults', txDefaults, json_schemas_1.schemas.txDataSchema, [
                            json_schemas_1.schemas.addressSchema,
                            json_schemas_1.schemas.numberSchema,
                            json_schemas_1.schemas.jsNumber,
                        ]);
                        provider = utils_1.providerUtils.standardizeOrThrow(supportedProvider);
                        constructorAbi = base_contract_1.BaseContract._lookupConstructorAbi(abi);
                        _a = __read(base_contract_1.BaseContract._formatABIDataItemList(constructorAbi.inputs, [_exchange, _zrxAssetData], base_contract_1.BaseContract._bigNumberToString), 2), _exchange = _a[0], _zrxAssetData = _a[1];
                        iface = new ethers.utils.Interface(abi);
                        deployInfo = iface.deployFunction;
                        txData = deployInfo.encode(bytecode, [_exchange, _zrxAssetData]);
                        web3Wrapper = new web3_wrapper_1.Web3Wrapper(provider);
                        return [4 /*yield*/, base_contract_1.BaseContract._applyDefaultsToTxDataAsync({ data: txData }, txDefaults, web3Wrapper.estimateGasAsync.bind(web3Wrapper))];
                    case 1:
                        txDataWithDefaults = _b.sent();
                        return [4 /*yield*/, web3Wrapper.sendTransactionAsync(txDataWithDefaults)];
                    case 2:
                        txHash = _b.sent();
                        utils_1.logUtils.log("transactionHash: " + txHash);
                        return [4 /*yield*/, web3Wrapper.awaitTransactionSuccessAsync(txHash)];
                    case 3:
                        txReceipt = _b.sent();
                        utils_1.logUtils.log("DevUtils successfully deployed at " + txReceipt.contractAddress);
                        contractInstance = new DevUtilsContract(txReceipt.contractAddress, provider, txDefaults, logDecodeDependencies);
                        contractInstance.constructorArgs = [_exchange, _zrxAssetData];
                        return [2 /*return*/, contractInstance];
                }
            });
        });
    };
    /**
     * @returns      The contract ABI
     */
    DevUtilsContract.ABI = function () {
        var abi = [
            {
                constant: true,
                inputs: [
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                name: 'decodeERC721AssetData',
                outputs: [
                    {
                        name: 'assetProxyId',
                        type: 'bytes4',
                    },
                    {
                        name: 'tokenAddress',
                        type: 'address',
                    },
                    {
                        name: 'tokenId',
                        type: 'uint256',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'ownerAddress',
                        type: 'address',
                    },
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                name: 'getBalanceAndAssetProxyAllowance',
                outputs: [
                    {
                        name: 'balance',
                        type: 'uint256',
                    },
                    {
                        name: 'allowance',
                        type: 'uint256',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [],
                name: 'ERC1155_PROXY_ID',
                outputs: [
                    {
                        name: '',
                        type: 'bytes4',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'ownerAddress',
                        type: 'address',
                    },
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                name: 'getTransferableAssetAmount',
                outputs: [
                    {
                        name: 'transferableAssetAmount',
                        type: 'uint256',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'ownerAddress',
                        type: 'address',
                    },
                    {
                        name: 'assetData',
                        type: 'bytes[]',
                    },
                ],
                name: 'getBatchAssetProxyAllowances',
                outputs: [
                    {
                        name: 'allowances',
                        type: 'uint256[]',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'tokenAddress',
                        type: 'address',
                    },
                ],
                name: 'encodeERC20AssetData',
                outputs: [
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'transactionData',
                        type: 'bytes',
                    },
                ],
                name: 'decodeZeroExTransactionData',
                outputs: [
                    {
                        name: 'functionName',
                        type: 'string',
                    },
                    {
                        name: 'orders',
                        type: 'tuple[]',
                        components: [
                            {
                                name: 'makerAddress',
                                type: 'address',
                            },
                            {
                                name: 'takerAddress',
                                type: 'address',
                            },
                            {
                                name: 'feeRecipientAddress',
                                type: 'address',
                            },
                            {
                                name: 'senderAddress',
                                type: 'address',
                            },
                            {
                                name: 'makerAssetAmount',
                                type: 'uint256',
                            },
                            {
                                name: 'takerAssetAmount',
                                type: 'uint256',
                            },
                            {
                                name: 'makerFee',
                                type: 'uint256',
                            },
                            {
                                name: 'takerFee',
                                type: 'uint256',
                            },
                            {
                                name: 'expirationTimeSeconds',
                                type: 'uint256',
                            },
                            {
                                name: 'salt',
                                type: 'uint256',
                            },
                            {
                                name: 'makerAssetData',
                                type: 'bytes',
                            },
                            {
                                name: 'takerAssetData',
                                type: 'bytes',
                            },
                        ],
                    },
                    {
                        name: 'takerAssetFillAmounts',
                        type: 'uint256[]',
                    },
                    {
                        name: 'signatures',
                        type: 'bytes[]',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'ownerAddress',
                        type: 'address',
                    },
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                name: 'getBalance',
                outputs: [
                    {
                        name: 'balance',
                        type: 'uint256',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'orders',
                        type: 'tuple[]',
                        components: [
                            {
                                name: 'makerAddress',
                                type: 'address',
                            },
                            {
                                name: 'takerAddress',
                                type: 'address',
                            },
                            {
                                name: 'feeRecipientAddress',
                                type: 'address',
                            },
                            {
                                name: 'senderAddress',
                                type: 'address',
                            },
                            {
                                name: 'makerAssetAmount',
                                type: 'uint256',
                            },
                            {
                                name: 'takerAssetAmount',
                                type: 'uint256',
                            },
                            {
                                name: 'makerFee',
                                type: 'uint256',
                            },
                            {
                                name: 'takerFee',
                                type: 'uint256',
                            },
                            {
                                name: 'expirationTimeSeconds',
                                type: 'uint256',
                            },
                            {
                                name: 'salt',
                                type: 'uint256',
                            },
                            {
                                name: 'makerAssetData',
                                type: 'bytes',
                            },
                            {
                                name: 'takerAssetData',
                                type: 'bytes',
                            },
                        ],
                    },
                    {
                        name: 'signatures',
                        type: 'bytes[]',
                    },
                ],
                name: 'getOrderRelevantStates',
                outputs: [
                    {
                        name: 'ordersInfo',
                        type: 'tuple[]',
                        components: [
                            {
                                name: 'orderStatus',
                                type: 'uint8',
                            },
                            {
                                name: 'orderHash',
                                type: 'bytes32',
                            },
                            {
                                name: 'orderTakerAssetFilledAmount',
                                type: 'uint256',
                            },
                        ],
                    },
                    {
                        name: 'fillableTakerAssetAmounts',
                        type: 'uint256[]',
                    },
                    {
                        name: 'isValidSignature',
                        type: 'bool[]',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [],
                name: 'ERC20_PROXY_ID',
                outputs: [
                    {
                        name: '',
                        type: 'bytes4',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                name: 'decodeERC20AssetData',
                outputs: [
                    {
                        name: 'assetProxyId',
                        type: 'bytes4',
                    },
                    {
                        name: 'tokenAddress',
                        type: 'address',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'order',
                        type: 'tuple',
                        components: [
                            {
                                name: 'makerAddress',
                                type: 'address',
                            },
                            {
                                name: 'takerAddress',
                                type: 'address',
                            },
                            {
                                name: 'feeRecipientAddress',
                                type: 'address',
                            },
                            {
                                name: 'senderAddress',
                                type: 'address',
                            },
                            {
                                name: 'makerAssetAmount',
                                type: 'uint256',
                            },
                            {
                                name: 'takerAssetAmount',
                                type: 'uint256',
                            },
                            {
                                name: 'makerFee',
                                type: 'uint256',
                            },
                            {
                                name: 'takerFee',
                                type: 'uint256',
                            },
                            {
                                name: 'expirationTimeSeconds',
                                type: 'uint256',
                            },
                            {
                                name: 'salt',
                                type: 'uint256',
                            },
                            {
                                name: 'makerAssetData',
                                type: 'bytes',
                            },
                            {
                                name: 'takerAssetData',
                                type: 'bytes',
                            },
                        ],
                    },
                    {
                        name: 'signature',
                        type: 'bytes',
                    },
                ],
                name: 'getOrderRelevantState',
                outputs: [
                    {
                        name: 'orderInfo',
                        type: 'tuple',
                        components: [
                            {
                                name: 'orderStatus',
                                type: 'uint8',
                            },
                            {
                                name: 'orderHash',
                                type: 'bytes32',
                            },
                            {
                                name: 'orderTakerAssetFilledAmount',
                                type: 'uint256',
                            },
                        ],
                    },
                    {
                        name: 'fillableTakerAssetAmount',
                        type: 'uint256',
                    },
                    {
                        name: 'isValidSignature',
                        type: 'bool',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                name: 'decodeERC1155AssetData',
                outputs: [
                    {
                        name: 'assetProxyId',
                        type: 'bytes4',
                    },
                    {
                        name: 'tokenAddress',
                        type: 'address',
                    },
                    {
                        name: 'tokenIds',
                        type: 'uint256[]',
                    },
                    {
                        name: 'tokenValues',
                        type: 'uint256[]',
                    },
                    {
                        name: 'callbackData',
                        type: 'bytes',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'addresses',
                        type: 'address[]',
                    },
                ],
                name: 'getEthBalances',
                outputs: [
                    {
                        name: '',
                        type: 'uint256[]',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [],
                name: 'ERC721_PROXY_ID',
                outputs: [
                    {
                        name: '',
                        type: 'bytes4',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'tokenAddress',
                        type: 'address',
                    },
                    {
                        name: 'tokenId',
                        type: 'uint256',
                    },
                ],
                name: 'encodeERC721AssetData',
                outputs: [
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [],
                name: 'MULTI_ASSET_PROXY_ID',
                outputs: [
                    {
                        name: '',
                        type: 'bytes4',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'tokenAddress',
                        type: 'address',
                    },
                    {
                        name: 'tokenIds',
                        type: 'uint256[]',
                    },
                    {
                        name: 'tokenValues',
                        type: 'uint256[]',
                    },
                    {
                        name: 'callbackData',
                        type: 'bytes',
                    },
                ],
                name: 'encodeERC1155AssetData',
                outputs: [
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'tokenAddress',
                        type: 'address',
                    },
                    {
                        name: 'tokenId',
                        type: 'uint256',
                    },
                ],
                name: 'getERC721TokenOwner',
                outputs: [
                    {
                        name: 'ownerAddress',
                        type: 'address',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                name: 'decodeMultiAssetData',
                outputs: [
                    {
                        name: 'assetProxyId',
                        type: 'bytes4',
                    },
                    {
                        name: 'amounts',
                        type: 'uint256[]',
                    },
                    {
                        name: 'nestedAssetData',
                        type: 'bytes[]',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'ownerAddress',
                        type: 'address',
                    },
                    {
                        name: 'assetData',
                        type: 'bytes[]',
                    },
                ],
                name: 'getBatchBalances',
                outputs: [
                    {
                        name: 'balances',
                        type: 'uint256[]',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'ownerAddress',
                        type: 'address',
                    },
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                name: 'getAssetProxyAllowance',
                outputs: [
                    {
                        name: 'allowance',
                        type: 'uint256',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'amounts',
                        type: 'uint256[]',
                    },
                    {
                        name: 'nestedAssetData',
                        type: 'bytes[]',
                    },
                ],
                name: 'encodeMultiAssetData',
                outputs: [
                    {
                        name: 'assetData',
                        type: 'bytes',
                    },
                ],
                payable: false,
                stateMutability: 'pure',
                type: 'function',
            },
            {
                constant: true,
                inputs: [],
                name: 'STATIC_CALL_PROXY_ID',
                outputs: [
                    {
                        name: '',
                        type: 'bytes4',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                constant: true,
                inputs: [
                    {
                        name: 'ownerAddress',
                        type: 'address',
                    },
                    {
                        name: 'assetData',
                        type: 'bytes[]',
                    },
                ],
                name: 'getBatchBalancesAndAssetProxyAllowances',
                outputs: [
                    {
                        name: 'balances',
                        type: 'uint256[]',
                    },
                    {
                        name: 'allowances',
                        type: 'uint256[]',
                    },
                ],
                payable: false,
                stateMutability: 'view',
                type: 'function',
            },
            {
                inputs: [
                    {
                        name: '_exchange',
                        type: 'address',
                    },
                    {
                        name: '_zrxAssetData',
                        type: 'bytes',
                    },
                ],
                outputs: [],
                payable: false,
                stateMutability: 'nonpayable',
                type: 'constructor',
            },
        ];
        return abi;
    };
    return DevUtilsContract;
}(base_contract_1.BaseContract));
exports.DevUtilsContract = DevUtilsContract;
// tslint:disable:max-file-line-count
// tslint:enable:no-unbound-method no-parameter-reassignment no-consecutive-blank-lines ordered-imports align
// tslint:enable:trailing-comma whitespace no-trailing-whitespace
