// @melonproject/melonjs/contracts/fund/participation/ParticipationFactory.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const createTestEnvironment_1 = require("../../../utils/tests/createTestEnvironment");
const randomAddress_1 = require("../../../utils/tests/randomAddress");
const deployHub_1 = require("../../../utils/tests/deployHub");
const deployWeth_1 = require("../../../utils/tests/deployWeth");
const deployRegistry_1 = require("../../../utils/tests/deployRegistry");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const ParticipationFactory_1 = require("./ParticipationFactory");
const ParticipationFactory_bin_1 = require("../../../abis/ParticipationFactory.bin");
describe('ParticipationFactory', () => {
    let environment;
    let participationFactory;
    beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
        environment = yield createTestEnvironment_1.createTestEnvironment();
        const deploy = ParticipationFactory_1.ParticipationFactory.deploy(environment, ParticipationFactory_bin_1.ParticipationFactoryBytecode, environment.accounts[0]);
        participationFactory = yield deploy.send(yield deploy.prepare());
    }));
    it('should check if a contract is an instance of the ParticipationFactory', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = yield participationFactory.isInstance(randomAddress_1.randomAddress());
        expect(typeof result).toBe('boolean');
    }));
    it('should create an instance of a Participation contract', () => __awaiter(void 0, void 0, void 0, function* () {
        const hub = yield deployHub_1.deployHub(environment, environment.accounts[0], {
            manager: environment.accounts[1],
            name: 'accounting-test-fund',
        });
        const weth = yield deployWeth_1.deployWeth(environment, environment.accounts[0]);
        const registry = yield deployRegistry_1.deployRegistry(environment, environment.accounts[0], environment.accounts[0]);
        const txRegisterAsset = yield registry.registerAsset(environment.accounts[0], {
            address: weth.contract.address,
            name: 'Test Asset',
            symbol: 'TAT',
            url: 'https://tat.tat/',
            reserveMin: new bignumber_js_1.default(100000),
            standards: [1, 2, 3],
            sigs: ['0x30303030'],
        });
        yield txRegisterAsset.send(yield txRegisterAsset.prepare());
        const tx = participationFactory.createInstance(environment.accounts[0], {
            hub: hub.contract.address,
            defaultAssets: [weth.contract.address],
            registry: registry.contract.address,
        });
        const txResult = yield tx.send(yield tx.prepare());
        expect(txResult.gasUsed).toBeGreaterThanOrEqual(0);
        expect(txResult.status).toBe(true);
    }));
});
