// @melonproject/melonjs/contracts/exchanges/third-party/oasisdex/OasisDexExchange.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const createTestEnvironment_1 = require("../../../../utils/tests/createTestEnvironment");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const OasisDexExchange_1 = require("./OasisDexExchange");
const OasisDexExchange_bin_1 = require("../../../../abis/OasisDexExchange.bin");
describe('OasisDex', () => {
    let environment;
    let oasisDex;
    beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
        environment = yield createTestEnvironment_1.createTestEnvironment();
        const deploy = OasisDexExchange_1.OasisDexExchange.deploy(environment, OasisDexExchange_bin_1.OasisDexExchangeBytecode, environment.accounts[0], new bignumber_js_1.default('999999999999'));
        oasisDex = yield deploy.send(yield deploy.prepare());
    }));
    it('should check if the exchange adapter has an address', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = oasisDex.contract.address;
        expect(result.startsWith('0x')).toBe(true);
    }));
    fit('should get an order', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = yield oasisDex.getOffer(new bignumber_js_1.default(0));
        expect(result).toMatchObject({
            makerQuantity: expect.any(bignumber_js_1.default),
            makerAsset: expect.any(String),
            takerQuantity: expect.any(bignumber_js_1.default),
            takerAsset: expect.any(String),
            owner: expect.any(String),
            timestamp: expect.any(Date),
        });
    }));
});
