// @melonproject/melonjs/contracts/fund/trading/exchanges/KyberTradingAdapter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const functionSignature_1 = require("../../../../utils/functionSignature");
const ExchangeAdapter_abi_1 = require("../../../../abis/ExchangeAdapter.abi");
const zeroAddress_1 = require("../../../../utils/zeroAddress");
const checkSufficientBalance_1 = require("../utils/checkSufficientBalance");
const checkFundIsNotShutdown_1 = require("../utils/checkFundIsNotShutdown");
const zeroBigNumber_1 = require("../../../../utils/zeroBigNumber");
const BaseTradingAdapter_1 = require("./BaseTradingAdapter");
const web3_utils_1 = require("web3-utils");
const checkSenderIsFundManager_1 = require("../utils/checkSenderIsFundManager");
class KyberTradingAdapter extends BaseTradingAdapter_1.BaseTradingAdapter {
    /**
     * Take order on Kyber
     *
     * @param from The address of the sender
     * @param args The arguments as [[KyberTakeOrderArgs]]
     */
    takeOrder(from, args) {
        const paddedZeros = web3_utils_1.padLeft('0x0', 64);
        const methodArgs = {
            exchangeIndex: this.index,
            methodSignature: functionSignature_1.functionSignature(ExchangeAdapter_abi_1.ExchangeAdapterAbi, 'takeOrder'),
            orderAddresses: [
                zeroAddress_1.zeroAddress,
                zeroAddress_1.zeroAddress,
                args.makerAsset,
                args.takerAsset,
                zeroAddress_1.zeroAddress,
                zeroAddress_1.zeroAddress,
                zeroAddress_1.zeroAddress,
                zeroAddress_1.zeroAddress,
            ],
            orderValues: [
                args.makerQuantity,
                args.takerQuantity,
                zeroBigNumber_1.zeroBigNumber,
                zeroBigNumber_1.zeroBigNumber,
                zeroBigNumber_1.zeroBigNumber,
                zeroBigNumber_1.zeroBigNumber,
                args.takerQuantity,
                zeroBigNumber_1.zeroBigNumber,
            ],
            orderData: [paddedZeros, paddedZeros, paddedZeros, paddedZeros],
            identifier: paddedZeros,
            signature: paddedZeros,
        };
        const validate = () => __awaiter(this, void 0, void 0, function* () {
            const vaultAddress = (yield this.trading.getRoutes()).vault;
            const hubAddress = yield this.trading.getHub();
            yield Promise.all([
                checkSufficientBalance_1.checkSufficientBalance(this.trading.environment, args.takerAsset, args.takerQuantity, vaultAddress),
                checkFundIsNotShutdown_1.checkFundIsNotShutdown(this.trading.environment, hubAddress),
                checkSenderIsFundManager_1.checkSenderIsFundManager(this.trading.environment, from, hubAddress),
            ]);
        });
        return this.trading.callOnExchange(from, methodArgs, validate);
    }
}
exports.KyberTradingAdapter = KyberTradingAdapter;
