// @melonproject/melonjs/contracts/fund/trading/TradingFactory.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const createTestEnvironment_1 = require("../../../utils/tests/createTestEnvironment");
const randomAddress_1 = require("../../../utils/tests/randomAddress");
const deployHub_1 = require("../../../utils/tests/deployHub");
const TradingFactory_1 = require("./TradingFactory");
const TradingFactory_bin_1 = require("../../../abis/TradingFactory.bin");
const deployRegistry_1 = require("../../../utils/tests/deployRegistry");
describe('TradingFactory', () => {
    let environment;
    let tradingFactory;
    beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
        environment = yield createTestEnvironment_1.createTestEnvironment();
        const deploy = TradingFactory_1.TradingFactory.deploy(environment, TradingFactory_bin_1.TradingFactoryBytecode, environment.accounts[0]);
        tradingFactory = yield deploy.send(yield deploy.prepare());
    }));
    it('should check if a contract is an instance of TradingFactory', () => __awaiter(void 0, void 0, void 0, function* () {
        const result = yield tradingFactory.isInstance(randomAddress_1.randomAddress());
        expect(typeof result).toBe('boolean');
    }));
    it('should create an instance of a trading contract', () => __awaiter(void 0, void 0, void 0, function* () {
        const hub = yield deployHub_1.deployHub(environment, environment.accounts[0], {
            manager: environment.accounts[1],
            name: 'trading-test-fund',
        });
        const exchangeAddress = randomAddress_1.randomAddress();
        const adapterAddress = randomAddress_1.randomAddress();
        const registry = yield deployRegistry_1.deployRegistry(environment, environment.accounts[0], environment.accounts[0]);
        const txExchangeAdapter = registry.registerExchangeAdapter(environment.accounts[0], {
            exchangeAddress,
            adapterAddress,
            takesCustody: true,
            sigs: ['0x30303030'],
        });
        yield txExchangeAdapter.send(yield txExchangeAdapter.prepare());
        const tx = tradingFactory.createInstance(environment.accounts[0], {
            hub: hub.contract.address,
            exchanges: [exchangeAddress],
            adapters: [adapterAddress],
            registry: registry.contract.address,
        });
        const txResult = yield tx.send(yield tx.prepare());
        expect(txResult.gasUsed).toBeGreaterThanOrEqual(0);
        expect(txResult.status).toBe(true);
    }));
});
