// @0x/abi-gen-wrappers/lib/src/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var asset_proxy_owner_1 = require("./generated-wrappers/asset_proxy_owner");
exports.AssetProxyOwnerEvents = asset_proxy_owner_1.AssetProxyOwnerEvents;
exports.AssetProxyOwnerContract = asset_proxy_owner_1.AssetProxyOwnerContract;
var dev_utils_1 = require("./generated-wrappers/dev_utils");
exports.DevUtilsContract = dev_utils_1.DevUtilsContract;
var dummy_erc20_token_1 = require("./generated-wrappers/dummy_erc20_token");
exports.DummyERC20TokenEvents = dummy_erc20_token_1.DummyERC20TokenEvents;
exports.DummyERC20TokenContract = dummy_erc20_token_1.DummyERC20TokenContract;
var dummy_erc721_token_1 = require("./generated-wrappers/dummy_erc721_token");
exports.DummyERC721TokenEvents = dummy_erc721_token_1.DummyERC721TokenEvents;
exports.DummyERC721TokenContract = dummy_erc721_token_1.DummyERC721TokenContract;
var dutch_auction_1 = require("./generated-wrappers/dutch_auction");
exports.DutchAuctionContract = dutch_auction_1.DutchAuctionContract;
var erc1155_proxy_1 = require("./generated-wrappers/erc1155_proxy");
exports.ERC1155ProxyEvents = erc1155_proxy_1.ERC1155ProxyEvents;
exports.ERC1155ProxyContract = erc1155_proxy_1.ERC1155ProxyContract;
var erc20_proxy_1 = require("./generated-wrappers/erc20_proxy");
exports.ERC20ProxyEvents = erc20_proxy_1.ERC20ProxyEvents;
exports.ERC20ProxyContract = erc20_proxy_1.ERC20ProxyContract;
var erc20_token_1 = require("./generated-wrappers/erc20_token");
exports.ERC20TokenEvents = erc20_token_1.ERC20TokenEvents;
exports.ERC20TokenContract = erc20_token_1.ERC20TokenContract;
var erc721_proxy_1 = require("./generated-wrappers/erc721_proxy");
exports.ERC721ProxyEvents = erc721_proxy_1.ERC721ProxyEvents;
exports.ERC721ProxyContract = erc721_proxy_1.ERC721ProxyContract;
var erc721_token_1 = require("./generated-wrappers/erc721_token");
exports.ERC721TokenEvents = erc721_token_1.ERC721TokenEvents;
exports.ERC721TokenContract = erc721_token_1.ERC721TokenContract;
var exchange_1 = require("./generated-wrappers/exchange");
exports.ExchangeEvents = exchange_1.ExchangeEvents;
exports.ExchangeContract = exchange_1.ExchangeContract;
var forwarder_1 = require("./generated-wrappers/forwarder");
exports.ForwarderContract = forwarder_1.ForwarderContract;
var i_asset_proxy_1 = require("./generated-wrappers/i_asset_proxy");
exports.IAssetProxyContract = i_asset_proxy_1.IAssetProxyContract;
var i_validator_1 = require("./generated-wrappers/i_validator");
exports.IValidatorContract = i_validator_1.IValidatorContract;
var i_wallet_1 = require("./generated-wrappers/i_wallet");
exports.IWalletContract = i_wallet_1.IWalletContract;
var multi_asset_proxy_1 = require("./generated-wrappers/multi_asset_proxy");
exports.MultiAssetProxyEvents = multi_asset_proxy_1.MultiAssetProxyEvents;
exports.MultiAssetProxyContract = multi_asset_proxy_1.MultiAssetProxyContract;
var order_validator_1 = require("./generated-wrappers/order_validator");
exports.OrderValidatorContract = order_validator_1.OrderValidatorContract;
var static_call_proxy_1 = require("./generated-wrappers/static_call_proxy");
exports.StaticCallProxyContract = static_call_proxy_1.StaticCallProxyContract;
var weth9_1 = require("./generated-wrappers/weth9");
exports.WETH9Events = weth9_1.WETH9Events;
exports.WETH9Contract = weth9_1.WETH9Contract;
var zrx_token_1 = require("./generated-wrappers/zrx_token");
exports.ZRXTokenEvents = zrx_token_1.ZRXTokenEvents;
exports.ZRXTokenContract = zrx_token_1.ZRXTokenContract;
var coordinator_1 = require("./generated-wrappers/coordinator");
exports.CoordinatorContract = coordinator_1.CoordinatorContract;
var coordinator_registry_1 = require("./generated-wrappers/coordinator_registry");
exports.CoordinatorRegistryEvents = coordinator_registry_1.CoordinatorRegistryEvents;
exports.CoordinatorRegistryContract = coordinator_registry_1.CoordinatorRegistryContract;
var eth_balance_checker_1 = require("./generated-wrappers/eth_balance_checker");
exports.EthBalanceCheckerContract = eth_balance_checker_1.EthBalanceCheckerContract;
__export(require("@0x/contract-addresses"));
