// @melonproject/melonjs/abis/AmguConsumer.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmguConsumerAbi = [
    {
        anonymous: false,
        inputs: [
            { indexed: true, internalType: 'address', name: 'payer', type: 'address' },
            { indexed: false, internalType: 'uint256', name: 'totalAmguPaidInEth', type: 'uint256' },
            { indexed: false, internalType: 'uint256', name: 'amguChargableGas', type: 'uint256' },
            { indexed: false, internalType: 'uint256', name: 'incentivePaid', type: 'uint256' },
        ],
        name: 'AmguPaid',
        type: 'event',
    },
    {
        inputs: [],
        name: 'engine',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'mlnToken',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'priceSource',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'registry',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
];
