// @melonproject/melonjs/abis/IAccounting.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IAccountingAbi = [
    {
        inputs: [{ internalType: 'address', name: 'ofAsset', type: 'address' }],
        name: 'calcAssetGAV',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'calcGav',
        outputs: [{ internalType: 'uint256', name: 'gav', type: 'uint256' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'calcGavPerShareNetManagementFee',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'gav', type: 'uint256' },
            { internalType: 'uint256', name: 'unclaimedFees', type: 'uint256' },
        ],
        name: 'calcNav',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'pure',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getFundHoldings',
        outputs: [
            { internalType: 'uint256[]', name: '', type: 'uint256[]' },
            { internalType: 'address[]', name: '', type: 'address[]' },
        ],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getOwnedAssetsLength',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'performCalculations',
        outputs: [
            { internalType: 'uint256', name: 'gav', type: 'uint256' },
            { internalType: 'uint256', name: 'unclaimedFees', type: 'uint256' },
            { internalType: 'uint256', name: 'feesInShares', type: 'uint256' },
            { internalType: 'uint256', name: 'nav', type: 'uint256' },
            { internalType: 'uint256', name: 'sharePrice', type: 'uint256' },
            { internalType: 'uint256', name: 'gavPerShareNetManagementFee', type: 'uint256' },
        ],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'totalValue', type: 'uint256' },
            { internalType: 'uint256', name: 'numShares', type: 'uint256' },
        ],
        name: 'valuePerShare',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
];
