// @melonproject/melonjs/abis/IParticipation.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IParticipationAbi = [
    { inputs: [], name: 'cancelRequest', outputs: [], stateMutability: 'payable', type: 'function' },
    {
        inputs: [{ internalType: 'address', name: 'requestOwner', type: 'address' }],
        name: 'executeRequestFor',
        outputs: [],
        stateMutability: 'payable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '', type: 'address' }],
        name: 'hasRequest',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    { inputs: [], name: 'redeem', outputs: [], stateMutability: 'nonpayable', type: 'function' },
    {
        inputs: [
            { internalType: 'uint256', name: 'shareQuantity', type: 'uint256' },
            { internalType: 'address[]', name: 'requestedAssets', type: 'address[]' },
        ],
        name: 'redeemWithConstraints',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'requestedShares', type: 'uint256' },
            { internalType: 'uint256', name: 'investmentAmount', type: 'uint256' },
            { internalType: 'address', name: 'investmentAsset', type: 'address' },
        ],
        name: 'requestInvestment',
        outputs: [],
        stateMutability: 'payable',
        type: 'function',
    },
];
