// @melonproject/melonjs/abis/IPriceSource.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IPriceSourceAbi = [
    {
        inputs: [
            { internalType: 'uint256', name: 'fromAssetQuantity', type: 'uint256' },
            { internalType: 'address', name: 'fromAsset', type: 'address' },
            { internalType: 'address', name: 'toAsset', type: 'address' },
        ],
        name: 'convertQuantity',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'sellAsset', type: 'address' },
            { internalType: 'address', name: 'buyAsset', type: 'address' },
        ],
        name: 'existsPriceOnAssetPair',
        outputs: [{ internalType: 'bool', name: 'isExistent', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: 'ofAsset', type: 'address' }],
        name: 'getInvertedPriceInfo',
        outputs: [
            { internalType: 'uint256', name: 'price', type: 'uint256' },
            { internalType: 'uint256', name: 'decimals', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getLastUpdate',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'sellAsset', type: 'address' },
            { internalType: 'uint256', name: 'sellQuantity', type: 'uint256' },
            { internalType: 'uint256', name: 'buyQuantity', type: 'uint256' },
        ],
        name: 'getOrderPriceInfo',
        outputs: [{ internalType: 'uint256', name: 'orderPrice', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '_asset', type: 'address' }],
        name: 'getPrice',
        outputs: [
            { internalType: 'uint256', name: 'price', type: 'uint256' },
            { internalType: 'uint256', name: 'timestamp', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '_asset', type: 'address' }],
        name: 'getPriceInfo',
        outputs: [
            { internalType: 'uint256', name: 'price', type: 'uint256' },
            { internalType: 'uint256', name: 'decimals', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address[]', name: '_assets', type: 'address[]' }],
        name: 'getPrices',
        outputs: [
            { internalType: 'uint256[]', name: 'prices', type: 'uint256[]' },
            { internalType: 'uint256[]', name: 'timestamps', type: 'uint256[]' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getQuoteAsset',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: '_base', type: 'address' },
            { internalType: 'address', name: '_quote', type: 'address' },
        ],
        name: 'getReferencePriceInfo',
        outputs: [
            { internalType: 'uint256', name: 'referencePrice', type: 'uint256' },
            { internalType: 'uint256', name: 'decimal', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '', type: 'address' }],
        name: 'hasValidPrice',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address[]', name: '', type: 'address[]' }],
        name: 'hasValidPrices',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
];
