// @melonproject/melonjs/abis/ITrading.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ITradingAbi = [
    {
        inputs: [
            { internalType: 'address', name: 'ofExchange', type: 'address' },
            { internalType: 'address', name: 'ofSellAsset', type: 'address' },
            { internalType: 'address', name: 'ofBuyAsset', type: 'address' },
            { internalType: 'address', name: 'ofFeeAsset', type: 'address' },
            { internalType: 'uint256', name: 'orderId', type: 'uint256' },
            { internalType: 'uint256', name: 'expiryTime', type: 'uint256' },
        ],
        name: 'addOpenMakeOrder',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'exchangeIndex', type: 'uint256' },
            { internalType: 'string', name: 'methodSignature', type: 'string' },
            { internalType: 'address[8]', name: 'orderAddresses', type: 'address[8]' },
            { internalType: 'uint256[8]', name: 'orderValues', type: 'uint256[8]' },
            { internalType: 'bytes[4]', name: 'orderData', type: 'bytes[4]' },
            { internalType: 'bytes32', name: 'identifier', type: 'bytes32' },
            { internalType: 'bytes', name: 'signature', type: 'bytes' },
        ],
        name: 'callOnExchange',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '_asset', type: 'address' }],
        name: 'getOpenMakeOrdersAgainstAsset',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'ofExchange', type: 'address' },
            { internalType: 'address', name: 'ofSellAsset', type: 'address' },
        ],
        name: 'removeOpenMakeOrder',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '_asset', type: 'address' }],
        name: 'updateAndGetQuantityBeingTraded',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
