// @melonproject/melonjs/abis/IUniswapExchange.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IUniswapExchangeAbi = [
    {
        inputs: [
            { internalType: 'uint256', name: 'min_liquidity', type: 'uint256' },
            { internalType: 'uint256', name: 'max_tokens', type: 'uint256' },
            { internalType: 'uint256', name: 'deadline', type: 'uint256' },
        ],
        name: 'addLiquidity',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'payable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'min_tokens', type: 'uint256' },
            { internalType: 'uint256', name: 'deadline', type: 'uint256' },
            { internalType: 'address', name: 'recipient', type: 'address' },
        ],
        name: 'ethToTokenTransferInput',
        outputs: [{ internalType: 'uint256', name: 'tokens_bought', type: 'uint256' }],
        stateMutability: 'payable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'uint256', name: 'eth_sold', type: 'uint256' }],
        name: 'getEthToTokenInputPrice',
        outputs: [{ internalType: 'uint256', name: 'tokens_bought', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'uint256', name: 'tokens_sold', type: 'uint256' }],
        name: 'getTokenToEthInputPrice',
        outputs: [{ internalType: 'uint256', name: 'eth_bought', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'tokens_sold', type: 'uint256' },
            { internalType: 'uint256', name: 'min_eth', type: 'uint256' },
            { internalType: 'uint256', name: 'deadline', type: 'uint256' },
        ],
        name: 'tokenToEthSwapInput',
        outputs: [{ internalType: 'uint256', name: 'eth_bought', type: 'uint256' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'tokens_sold', type: 'uint256' },
            { internalType: 'uint256', name: 'min_tokens_bought', type: 'uint256' },
            { internalType: 'uint256', name: 'min_eth_bought', type: 'uint256' },
            { internalType: 'uint256', name: 'deadline', type: 'uint256' },
            { internalType: 'address', name: 'recipient', type: 'address' },
            { internalType: 'address', name: 'token_addr', type: 'address' },
        ],
        name: 'tokenToTokenTransferInput',
        outputs: [{ internalType: 'uint256', name: 'tokens_bought', type: 'uint256' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
