// @melonproject/melonjs/abis/IWrapperLock.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IWrapperLockAbi = [
    {
        inputs: [{ internalType: 'address', name: '', type: 'address' }],
        name: 'balanceOf',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: '', type: 'uint256' },
            { internalType: 'uint256', name: '', type: 'uint256' },
        ],
        name: 'deposit',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: '', type: 'uint256' },
            { internalType: 'uint8', name: '', type: 'uint8' },
            { internalType: 'bytes32', name: '', type: 'bytes32' },
            { internalType: 'bytes32', name: '', type: 'bytes32' },
            { internalType: 'uint256', name: '', type: 'uint256' },
        ],
        name: 'withdraw',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
