// @melonproject/melonjs/abis/IZeroExV3.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IZeroExV3Abi = [
    {
        inputs: [
            {
                components: [
                    { internalType: 'address', name: 'makerAddress', type: 'address' },
                    { internalType: 'address', name: 'takerAddress', type: 'address' },
                    { internalType: 'address', name: 'feeRecipientAddress', type: 'address' },
                    { internalType: 'address', name: 'senderAddress', type: 'address' },
                    { internalType: 'uint256', name: 'makerAssetAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerAssetAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'makerFee', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerFee', type: 'uint256' },
                    { internalType: 'uint256', name: 'expirationTimeSeconds', type: 'uint256' },
                    { internalType: 'uint256', name: 'salt', type: 'uint256' },
                    { internalType: 'bytes', name: 'makerAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'takerAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'makerFeeAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'takerFeeAssetData', type: 'bytes' },
                ],
                internalType: 'struct IZeroExV3.Order',
                name: '',
                type: 'tuple',
            },
        ],
        name: 'cancelOrder',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'bytes32', name: '', type: 'bytes32' }],
        name: 'cancelled',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                components: [
                    { internalType: 'address', name: 'makerAddress', type: 'address' },
                    { internalType: 'address', name: 'takerAddress', type: 'address' },
                    { internalType: 'address', name: 'feeRecipientAddress', type: 'address' },
                    { internalType: 'address', name: 'senderAddress', type: 'address' },
                    { internalType: 'uint256', name: 'makerAssetAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerAssetAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'makerFee', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerFee', type: 'uint256' },
                    { internalType: 'uint256', name: 'expirationTimeSeconds', type: 'uint256' },
                    { internalType: 'uint256', name: 'salt', type: 'uint256' },
                    { internalType: 'bytes', name: 'makerAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'takerAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'makerFeeAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'takerFeeAssetData', type: 'bytes' },
                ],
                internalType: 'struct IZeroExV3.Order',
                name: '',
                type: 'tuple',
            },
            { internalType: 'uint256', name: '', type: 'uint256' },
            { internalType: 'bytes', name: '', type: 'bytes' },
        ],
        name: 'fillOrder',
        outputs: [
            {
                components: [
                    { internalType: 'uint256', name: 'makerAssetFilledAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerAssetFilledAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'makerFeePaid', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerFeePaid', type: 'uint256' },
                    { internalType: 'uint256', name: 'protocolFeePaid', type: 'uint256' },
                ],
                internalType: 'struct IZeroExV3.FillResults',
                name: '',
                type: 'tuple',
            },
        ],
        stateMutability: 'payable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'bytes32', name: '', type: 'bytes32' }],
        name: 'filled',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'bytes4', name: '', type: 'bytes4' }],
        name: 'getAssetProxy',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                components: [
                    { internalType: 'address', name: 'makerAddress', type: 'address' },
                    { internalType: 'address', name: 'takerAddress', type: 'address' },
                    { internalType: 'address', name: 'feeRecipientAddress', type: 'address' },
                    { internalType: 'address', name: 'senderAddress', type: 'address' },
                    { internalType: 'uint256', name: 'makerAssetAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerAssetAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'makerFee', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerFee', type: 'uint256' },
                    { internalType: 'uint256', name: 'expirationTimeSeconds', type: 'uint256' },
                    { internalType: 'uint256', name: 'salt', type: 'uint256' },
                    { internalType: 'bytes', name: 'makerAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'takerAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'makerFeeAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'takerFeeAssetData', type: 'bytes' },
                ],
                internalType: 'struct IZeroExV3.Order',
                name: '',
                type: 'tuple',
            },
        ],
        name: 'getOrderInfo',
        outputs: [
            {
                components: [
                    { internalType: 'uint8', name: 'orderStatus', type: 'uint8' },
                    { internalType: 'bytes32', name: 'orderHash', type: 'bytes32' },
                    { internalType: 'uint256', name: 'orderTakerAssetFilledAmount', type: 'uint256' },
                ],
                internalType: 'struct IZeroExV3.OrderInfo',
                name: '',
                type: 'tuple',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                components: [
                    { internalType: 'address', name: 'makerAddress', type: 'address' },
                    { internalType: 'address', name: 'takerAddress', type: 'address' },
                    { internalType: 'address', name: 'feeRecipientAddress', type: 'address' },
                    { internalType: 'address', name: 'senderAddress', type: 'address' },
                    { internalType: 'uint256', name: 'makerAssetAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerAssetAmount', type: 'uint256' },
                    { internalType: 'uint256', name: 'makerFee', type: 'uint256' },
                    { internalType: 'uint256', name: 'takerFee', type: 'uint256' },
                    { internalType: 'uint256', name: 'expirationTimeSeconds', type: 'uint256' },
                    { internalType: 'uint256', name: 'salt', type: 'uint256' },
                    { internalType: 'bytes', name: 'makerAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'takerAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'makerFeeAssetData', type: 'bytes' },
                    { internalType: 'bytes', name: 'takerFeeAssetData', type: 'bytes' },
                ],
                internalType: 'struct IZeroExV3.Order',
                name: '',
                type: 'tuple',
            },
            { internalType: 'bytes', name: '', type: 'bytes' },
        ],
        name: 'isValidOrderSignature',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'bytes32', name: '', type: 'bytes32' }],
        name: 'preSign',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'protocolFeeCollector',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'protocolFeeMultiplier',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
];
