// @melonproject/melonjs/abis/MockVersion.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockVersionAbi = [
    {
        inputs: [],
        name: 'amguPrice',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getAmguPrice',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getShutDownStatus',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'isShutDown',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    { inputs: [], name: 'securityShutDown', outputs: [], stateMutability: 'nonpayable', type: 'function' },
    {
        inputs: [{ internalType: 'uint256', name: '_price', type: 'uint256' }],
        name: 'setAmguPrice',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '_hub', type: 'address' }],
        name: 'shutDownFund',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
