// @melonproject/melonjs/abis/MockFee.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockFeeAbi = [
    {
        inputs: [{ internalType: 'uint256', name: '_feeNumber', type: 'uint256' }],
        stateMutability: 'nonpayable',
        type: 'constructor',
    },
    {
        inputs: [],
        name: 'FEE_PERIOD',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'FEE_RATE',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'fee',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'feeAmount',
        outputs: [{ internalType: 'uint256', name: 'feeInShares', type: 'uint256' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'feeNumber',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'identifier',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'feeRate', type: 'uint256' },
            { internalType: 'uint256', name: 'feePeriod', type: 'uint256' },
            { internalType: 'address', name: 'denominationAsset', type: 'address' },
        ],
        name: 'initializeForUser',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'uint256', name: 'amount', type: 'uint256' }],
        name: 'setFeeAmount',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    { inputs: [], name: 'updateState', outputs: [], stateMutability: 'nonpayable', type: 'function' },
];
