// @melonproject/melonjs/abis/MockHub.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockHubAbi = [
    {
        anonymous: false,
        inputs: [
            { indexed: true, internalType: 'bytes32', name: 'src', type: 'bytes32' },
            { indexed: true, internalType: 'bytes32', name: 'dst', type: 'bytes32' },
            { indexed: true, internalType: 'bytes32', name: 'sig', type: 'bytes32' },
        ],
        name: 'LogForbid',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [
            { indexed: true, internalType: 'bytes32', name: 'src', type: 'bytes32' },
            { indexed: true, internalType: 'bytes32', name: 'dst', type: 'bytes32' },
            { indexed: true, internalType: 'bytes32', name: 'sig', type: 'bytes32' },
        ],
        name: 'LogPermit',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [{ indexed: true, internalType: 'address', name: 'authority', type: 'address' }],
        name: 'LogSetAuthority',
        type: 'event',
    },
    {
        anonymous: false,
        inputs: [{ indexed: true, internalType: 'address', name: 'owner', type: 'address' }],
        name: 'LogSetOwner',
        type: 'event',
    },
    {
        inputs: [],
        name: 'ANY',
        outputs: [{ internalType: 'bytes32', name: '', type: 'bytes32' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'accounting',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'authority',
        outputs: [{ internalType: 'contract DSAuthority', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'src_', type: 'address' },
            { internalType: 'address', name: 'dst_', type: 'address' },
            { internalType: 'bytes4', name: 'sig', type: 'bytes4' },
        ],
        name: 'canCall',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'src', type: 'address' },
            { internalType: 'address', name: 'dst', type: 'address' },
            { internalType: 'bytes32', name: 'sig', type: 'bytes32' },
        ],
        name: 'forbid',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'bytes32', name: 'src', type: 'bytes32' },
            { internalType: 'bytes32', name: 'dst', type: 'bytes32' },
            { internalType: 'bytes32', name: 'sig', type: 'bytes32' },
        ],
        name: 'forbid',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '_spoke', type: 'address' }],
        name: 'initializeSpoke',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'isShutDown',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'manager',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'name',
        outputs: [{ internalType: 'string', name: '', type: 'string' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'owner',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'participation',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'src', type: 'address' },
            { internalType: 'address', name: 'dst', type: 'address' },
            { internalType: 'bytes32', name: 'sig', type: 'bytes32' },
        ],
        name: 'permit',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'bytes32', name: 'src', type: 'bytes32' },
            { internalType: 'bytes32', name: 'dst', type: 'bytes32' },
            { internalType: 'bytes32', name: 'sig', type: 'bytes32' },
        ],
        name: 'permit',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: '_from', type: 'address' },
            { internalType: 'address', name: '_to', type: 'address' },
            { internalType: 'bytes4', name: '_sig', type: 'bytes4' },
        ],
        name: 'permitSomething',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'policyManager',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'priceSource',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'registry',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'routes',
        outputs: [
            { internalType: 'address', name: 'accounting', type: 'address' },
            { internalType: 'address', name: 'feeManager', type: 'address' },
            { internalType: 'address', name: 'participation', type: 'address' },
            { internalType: 'address', name: 'policyManager', type: 'address' },
            { internalType: 'address', name: 'shares', type: 'address' },
            { internalType: 'address', name: 'trading', type: 'address' },
            { internalType: 'address', name: 'vault', type: 'address' },
            { internalType: 'address', name: 'registry', type: 'address' },
            { internalType: 'address', name: 'version', type: 'address' },
            { internalType: 'address', name: 'engine', type: 'address' },
            { internalType: 'address', name: 'mlnAddress', type: 'address' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'contract DSAuthority', name: 'authority_', type: 'address' }],
        name: 'setAuthority',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '_manager', type: 'address' }],
        name: 'setManager',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'string', name: '_name', type: 'string' }],
        name: 'setName',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: 'owner_', type: 'address' }],
        name: 'setOwner',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    { inputs: [], name: 'setPermissions', outputs: [], stateMutability: 'nonpayable', type: 'function' },
    { inputs: [], name: 'setRouting', outputs: [], stateMutability: 'nonpayable', type: 'function' },
    {
        inputs: [{ internalType: 'bool', name: '_state', type: 'bool' }],
        name: 'setShutDownState',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address[11]', name: '_spokes', type: 'address[11]' }],
        name: 'setSpokes',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'shares',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    { inputs: [], name: 'shutDownFund', outputs: [], stateMutability: 'nonpayable', type: 'function' },
    {
        inputs: [],
        name: 'trading',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'vault',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
];
