// @melonproject/melonjs/abis/TestingPriceFeed.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestingPriceFeedAbi = [
    {
        inputs: [
            { internalType: 'address', name: '_quoteAsset', type: 'address' },
            { internalType: 'uint256', name: '_quoteDecimals', type: 'uint256' },
        ],
        stateMutability: 'nonpayable',
        type: 'constructor',
    },
    {
        anonymous: false,
        inputs: [
            { indexed: false, internalType: 'address[]', name: 'token', type: 'address[]' },
            { indexed: false, internalType: 'uint256[]', name: 'price', type: 'uint256[]' },
        ],
        name: 'PriceUpdate',
        type: 'event',
    },
    {
        inputs: [],
        name: 'QUOTE_ASSET',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '', type: 'address' }],
        name: 'assetsToDecimals',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '', type: 'address' }],
        name: 'assetsToPrices',
        outputs: [
            { internalType: 'uint256', name: 'price', type: 'uint256' },
            { internalType: 'uint256', name: 'timestamp', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address[]', name: '_assets', type: 'address[]' },
            { internalType: 'uint256[]', name: '_decimals', type: 'uint256[]' },
        ],
        name: 'batchSetDecimals',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'uint256', name: 'fromAssetQuantity', type: 'uint256' },
            { internalType: 'address', name: 'fromAsset', type: 'address' },
            { internalType: 'address', name: 'toAsset', type: 'address' },
        ],
        name: 'convertQuantity',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'sellAsset', type: 'address' },
            { internalType: 'address', name: 'buyAsset', type: 'address' },
        ],
        name: 'existsPriceOnAssetPair',
        outputs: [{ internalType: 'bool', name: 'isExistent', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: 'ofAsset', type: 'address' }],
        name: 'getInvertedPriceInfo',
        outputs: [
            { internalType: 'uint256', name: 'invertedPrice', type: 'uint256' },
            { internalType: 'uint256', name: 'assetDecimals', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getLastUpdate',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getLastUpdateId',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'sellAsset', type: 'address' },
            { internalType: 'uint256', name: 'sellQuantity', type: 'uint256' },
            { internalType: 'uint256', name: 'buyQuantity', type: 'uint256' },
        ],
        name: 'getOrderPriceInfo',
        outputs: [{ internalType: 'uint256', name: 'orderPrice', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: 'ofAsset', type: 'address' }],
        name: 'getPrice',
        outputs: [
            { internalType: 'uint256', name: 'price', type: 'uint256' },
            { internalType: 'uint256', name: 'timestamp', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: 'ofAsset', type: 'address' }],
        name: 'getPriceInfo',
        outputs: [
            { internalType: 'uint256', name: 'price', type: 'uint256' },
            { internalType: 'uint256', name: 'assetDecimals', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address[]', name: 'ofAssets', type: 'address[]' }],
        name: 'getPrices',
        outputs: [
            { internalType: 'uint256[]', name: '', type: 'uint256[]' },
            { internalType: 'uint256[]', name: '', type: 'uint256[]' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'getQuoteAsset',
        outputs: [{ internalType: 'address', name: '', type: 'address' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: 'ofBase', type: 'address' },
            { internalType: 'address', name: 'ofQuote', type: 'address' },
        ],
        name: 'getReferencePriceInfo',
        outputs: [
            { internalType: 'uint256', name: 'referencePrice', type: 'uint256' },
            { internalType: 'uint256', name: 'decimal', type: 'uint256' },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address', name: '_asset', type: 'address' }],
        name: 'hasValidPrice',
        outputs: [{ internalType: 'bool', name: 'isValid', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'address[]', name: '_assets', type: 'address[]' }],
        name: 'hasValidPrices',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [],
        name: 'lastUpdate',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address', name: '_asset', type: 'address' },
            { internalType: 'uint256', name: '_decimal', type: 'uint256' },
        ],
        name: 'setDecimals',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'bool', name: '_state', type: 'bool' }],
        name: 'setIsRecent',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'bool', name: '_state', type: 'bool' }],
        name: 'setNeverValid',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'address[]', name: '_assets', type: 'address[]' },
            { internalType: 'uint256[]', name: '_prices', type: 'uint256[]' },
        ],
        name: 'update',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [],
        name: 'updateId',
        outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
        stateMutability: 'view',
        type: 'function',
    },
];
