// @melonproject/melonjs/abis/TruePolicy.abi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TruePolicyAbi = [
    { inputs: [], stateMutability: 'nonpayable', type: 'constructor' },
    {
        inputs: [],
        name: 'identifier',
        outputs: [{ internalType: 'string', name: '', type: 'string' }],
        stateMutability: 'pure',
        type: 'function',
    },
    {
        inputs: [],
        name: 'position',
        outputs: [{ internalType: 'enum BooleanPolicy.Applied', name: '', type: 'uint8' }],
        stateMutability: 'pure',
        type: 'function',
    },
    {
        inputs: [
            { internalType: 'bytes4', name: 'sig', type: 'bytes4' },
            { internalType: 'address[5]', name: 'addresses', type: 'address[5]' },
            { internalType: 'uint256[3]', name: 'values', type: 'uint256[3]' },
            { internalType: 'bytes32', name: 'identifier', type: 'bytes32' },
        ],
        name: 'rule',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
